### DISCUSSION TITLE
Where are Closed Alerts Stored for CodeQL?

### DISCUSSION THREAD

**Discussion by ArielSAdamsNASA** [community/community#23399] [New to GitHub]
<p>How does CodeQL know which queries are closed between analyses?</p>
<p>Where are the closed alerts stored? If they are in the SARIF file, what do they look like?<br>
I tried to find the tag “Closed” in a SARIF file, but found no results.</p>
<p>Is there a way to keep closed alerts between a main and forked repository?<br>
When cloning a repository and running CodeQL, the closed alerts are not transferred from the main repo.</p>

**Comment by hmakholm** [comment, marked as answer]
<p>Strictly speaking, <strong>CodeQL</strong> does not know about closed alerts at all. It’s just an analysis tool that takes a particular source tree and produces a list of alerts that are <em>present in that source tree</em>. The alert list, as a SARIF file, is then uploaded to GitHub for further processing by a different component known as <strong>Code Scanning</strong>. It is Code Scanning that compares results from different commits in a repository to decide that some alerts are “the same” between them, and others are “new” or “fixed”.</p>
<p>This might seem like useless pedantry, but the difference in naming is relevant for your question because <em>CodeQL is the part you can see run</em>. If you’re using the default Actions setup, you can see the mumblings of the CodeQL engine in the log tabs of your workflow run; if you’re an enterprise customer running your own CI infrastructure, you’ll have scripts that execute CodeQL directly.</p>
<p>However, after the web backend has received the SARIF file, the internals of what its Code Scanning component does are not directly visible to users. Alerts for each repository and commit are stored in an internal database. We provide <a href="https://docs.github.com/en/rest/reference/code-scanning">some REST APIs</a> that interact with it, but they’re not a direct view of the database, and except for SARIF uploads they mostly work on a level of abstraction where matching of alerts between different commits in a repo has already happened.</p>
<p>We’re aware that the Code Scanning experience with forked repositories is not very good. We’re working towards improvements here, but there are both technical and policy challenges to solve along the way, as well as competing demands on the relevant engineering teams, so some patience will be necessary.</p>
